src/error/mod.rs

```diff
@@ -487,6 +488,22 @@ impl FederationError {
     }
 }
 
+// Converts a SingleFederationError array into `Result<(), FederationError>`.
+// The return value can be either Ok, Err with a SingleFederationError or MultipleFederationErrors,
+// depending on the number of errors in the input.
+pub(crate) fn check_federation_errors(
```

Comment:
I feel like `check_federation_errors` isn't doing a check, but more of a conversion 🤔 . For these sorts of things, we usually like a [`Into`](https://doc.rust-lang.org/std/convert/trait.Into.html) or [`TryInto`](https://doc.rust-lang.org/std/convert/trait.TryInto.html). 

What do you think about something like `pub(crate) struct SingleFederationErrorList { inner: Vec<SingleFederationError> }` (and pub(crate) so it's just internal). We already have a [`DiagnosticList` above that does a similar conversion](https://github.com/apollographql/federation-next/pull/251/files#diff-f43f218cc5fee24623f17725c9bc2c541d64b2e9c1d8ca5345de67ef2a3b7392R469-R474). That way [in the field_set.rs file](https://github.com/apollographql/federation-next/pull/251/files#diff-87a8f5443445f15c22bbc48ca16e97c3d26ac790c690ee10ed2d4ed4620418f7R27) you can just do an `alias_errors.into()` as your return

Reply:
I agree `check_federation_errors` is misleading. It's indeed a conversion function.

Originally, I wanted to use `Vec<SingleFederationErrors>` to collect errors, which includes the `Ok` case when empty. We sometimes use `MultipleFederationErrors` as both `Ok`(when the internal array is empty) and `Err` cases temporarily. But, I think `Vec` is clearer since `MultipleFederationErrors` will be cast into `Err` by default (so you have to check before returning it).

In order to convert `Vec<SingleFederationErrors>` into `Result`, I tried something like `From<&[SingleFederationError]> for Result<(), FederationError>`. But, that wasn't allowed. That's why I ended up with `check_federation_errors` function.

Now I think of this. I just got a better idea. While I can't convert `Vec<SingleFederationErrors>` using `into()`, since I have `FromIterator<SingleFederationErrors>`, I can do this:

```rust
impl FromIterator<SingleFederationError> for Result<(), FederationError> {
    fn from_iter<T: IntoIterator<Item=SingleFederationError>>(iter: T) -> Self {
        let errors: Vec<_> = iter.into_iter().collect();
        match errors.len().cmp(&1) {
            Ordering::Less => Ok(()),
            Ordering::Equal => Err(errors[0].clone().into()),
            Ordering::Greater => Err(MultipleFederationErrors {
                errors: errors.to_vec(),
            }
            .into()),
        }
    }
}
```
I'll push this version.

Reply:
@lrlna I pushed a refactored version. Please review :)

Reply:
I think we can simplify this a bit. The `IntoIterator` is doing a bit more than it should here. It looks like we want to have two things: a `Vec<SingleFederationError>` to be converted to `FederationError` to then be converted to `Result<(), FederationError>`.

The most rust-y way of doing it is doing what we do with `DiagnosticList` in `apollo-compiler`. 

a `Vec<Diagnostic>` gets converted to `DiagnosticList` (with an `Into` impl), a `DiagnosticList` has an impl with an `into_result` implementation (that returns `Result<(), Self>`, but you can probably have `Result<(), FederationError>).